libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/NavigationHelper.kt

```diff
@@ -0,0 +1,184 @@
+    @JvmStatic
+    fun routeDistanceRemaining(
+        legDistanceRemaining: Double,
+        legIndex: Int,
+        directionsRoute: DirectionsRoute
+    ): Double {
+        var legDistanceRemaining = legDistanceRemaining
+        directionsRoute.legs()?.let { legs ->
+            val legsSize = legs.size
+            if (legsSize < 2) {
+                return legDistanceRemaining
+            }
+
+            for (i in legIndex + 1 until legsSize) {
+                legDistanceRemaining += legs[i].distance() ?: 0.0
+            }
+        }
+        return legDistanceRemaining
+    }
+
+    /**
+     * Given the current [DirectionsRoute] and leg / step index,
+     * return a list of [Point] representing the current step.
+     *
+     *
+     * This method is only used on a per-step basis as [PolylineUtils.decode]
+     * can be a heavy operation based on the length of the step.
+     *
+     *
+     * Returns null if index is invalid.
+     *
+     * @param directionsRoute for list of steps
+     * @param legIndex to get current step list
+     * @param stepIndex to get current step
+     * @return list of [Point] representing the current step
+     */
+
+    @JvmStatic
+    fun decodeStepPoints(
+        directionsRoute: DirectionsRoute,
+        currentPoints: List<Point>?,
+        legIndex: Int,
+        stepIndex: Int
+    ): List<Point>? {
+        val legs = directionsRoute.legs()
+        legs?.let { legs ->
+            if (legs.isEmpty()) return currentPoints
+            val steps = legs[legIndex].steps()
+            steps?.let { steps ->
+                if (steps.isEmpty()) return currentPoints
+                val invalidStepIndex = stepIndex < 0 || stepIndex > steps.size - 1
+                if (invalidStepIndex) return currentPoints
+                val step = steps[stepIndex]
+                step?.let { step ->
+                    val stepGeometry = step.geometry()
+                    stepGeometry?.let {
+                        return PolylineUtils.decode(stepGeometry, PRECISION_6)
+                    }
+                } ?: return currentPoints
+            }
+        }
+        return currentPoints
+    }
+
+    /**
+     * Given a list of distance annotations, find the current annotation index.  This index retrieves the
+     * current annotation from any provided annotation list in [LegAnnotation].
+     *
+     * @param currentLegAnnotation current annotation being traveled along
+     * @param leg holding each list of annotations
+     * @param legDistanceRemaining to determine the new set of annotations
+     * @return a current set of annotation data for the user's position along the route
+     */
+    @JvmStatic
+    fun createCurrentAnnotation(
+        currentLegAnnotation: CurrentLegAnnotation?,
+        leg: RouteLeg,
+        legDistanceRemaining: Double
+    ): CurrentLegAnnotation? {
+        leg.annotation()?.let { legAnnotation ->
+            legAnnotation.distance()?.let { distanceList ->
```

Comment:
Could we join it like: 
```suggestion
            leg.annotation()?.distance()?.let { distanceList ->
```
??